packages/edit-post/src/components/layout/index.js

```diff
@@ -250,19 +331,59 @@ function MetaBoxesMain( { isLegacy } ) {
 				target.setPointerCapture( pointerId );
 			} }
 			onResizeStart={ ( event, direction, elementRef ) => {
-				// Avoids height jumping in case it’s limited by max-height.
-				elementRef.style.height = `${ elementRef.offsetHeight }px`;
-				// Stops initial max-height from being applied.
-				elementRef.classList.add( 'has-user-size' );
+				if ( isAutoHeight ) {
+					const heightNow = elementRef.offsetHeight;
+					// Sets the starting height to avoid visual jumps in height and
+					// aria-valuenow being `NaN` for the first (few) resize events.
+					resizableBoxRef.current.updateSize( { height: heightNow } );
+					// Causes `maxHeight` to update to full `max` value instead of half.
+					setIsUntouched( false );
+				}
+			} }
+			onResize={ ( event, direction, elementRef, delta ) => {
+				const fromHeight = resizableBoxRef.current.state.height;
+				const nextHeight = fromHeight + delta.height;
+				const separator = separatorRef.current;
+				separator.ariaValueNow = getAriaValueNow( nextHeight, max );
 			} }
 			onResizeStop={ () => {
+				const nextHeight = resizableBoxRef.current.state.height;
 				setPreference(
 					'core/edit-post',
 					'metaBoxesMainOpenHeight',
-					resizableBoxRef.current.state.height
+					nextHeight
 				);
 			} }
+			handleClasses={ {
+				top: 'edit-post-meta-boxes-main__resize-handle',
+			} }
+			handleComponent={ {
+				top: (
+					<>
+						<Tooltip text={ __( 'Drag to resize' ) }>
+							{ /* Disable reason: aria-valuenow is supported by separator role. */ }
+							{ /* eslint-disable-next-line jsx-a11y/role-supports-aria-props */ }
+							<button
+								ref={ separatorRef }
+								aria-label={ __( 'Drag to resize' ) }
+								aria-describedby={ separatorHelpId }
+								onKeyDown={ onSeparatorKeyDown }
+								// Disable reason: buttons are allowed to be separator role.
+								// eslint-disable-next-line jsx-a11y/no-interactive-element-to-noninteractive-role
+								role="separator"
+								aria-valuenow={ usedAriaValueNow }
+							/>
+						</Tooltip>
+						<VisuallyHidden id={ separatorHelpId }>
+							{ __(
+								'Use up and down arrow keys to resize the metabox panel.'
+							) }
+						</VisuallyHidden>
+					</>
+				),
+			} }
 		>
+			<meta ref={ effectSizeContraints } />
```

Comment:
I didn't know there was an approach like this. Couldn't we just use `useResizeObserver` + `useEffect` instead?

Reply:
I’ve used the approach before but it’s not something I’ve seen around. Here it was just an alternative to merging refs since this ref effect doesn’t require attaching to a specific element (just one that’s a child of the interface skeleton content) and other good candidates are already ref’d. `useEffect` could be used but would still need an element for reference and I like that it comes with `useRefEffect`.

I avoided `useResizeObserver` because the size needed here is a derivative of three elements and the elements are outside the component. I speculate that the hook would be less a little less efficient (creating three instances of `ResizeObserver`) and its usage more verbose/indirect due to a need to manually call fire ref callbacks after querying to find their intended elements (and how’s that for unidiomatic). Also, digressing a bit, I first came up with this just before #64820 appeared. Before it landed `useResizeObserver` would have been extremely hacky to try using since it needed to have an element inserted into the one to be observed.

Aside: just noticed I typo’d "constraints" in that name.